subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -38,40 +266,83 @@ static int download_client_callback(const struct download_client_evt *event)
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return err;
 		}
+
 		if (size > PM_MCUBOOT_SECONDARY_SIZE) {
 			LOG_ERR("Requested file too big to fit in flash\n");
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return -EFBIG;
 		}
 
-		err = flash_img_buffered_write(&flash_img,
-				(u8_t *)event->fragment.buf,
-				event->fragment.len, false);
-		if (err != 0) {
-			LOG_ERR("flash_img_buffered_write error %d", err);
+		if (first_fragment) {
+			first_fragment = false;
+			image_type = check_img_type((u8_t *) event->fragment.buf
+						    , event->fragment.len);
+		}
+
+		if (image_type == MCUBOOT) {
+			err = mcuboot_write(&flash_img,
+					    (u8_t *)event->fragment.buf,
+					    event->fragment.len);
+			if (err != 0) {
+				LOG_ERR("flash_img_buffered_write error %d",
+					err);
+				callback(FOTA_DOWNLOAD_EVT_ERROR);
+				return err;
+			}
+		}
+#ifdef CONFIG_FOTA_DOWNLOAD_MODEM_FW_UPDATE_SUPPORT
+		else if (image_type == MODEM_FW) {
+			err = modem_fw_write((u8_t *)event->fragment.buf,
+						     event->fragment.len);
+
+			if (err != 0) {
+				LOG_ERR("modem update error: %d", err);
+				err = download_client_disconnect(&dfu);
+				if (err) {
+					LOG_ERR("Failed to close socket!\n");
+				}
+				LOG_INF("Socket closed\n");
+				callback(FOTA_DOWNLOAD_EVT_ERROR);
+				return err;
+			}
+			LOG_INF("Wrote modem firmware");
+		}
+#endif
+		else {
+			LOG_ERR("Unknown image_type: %d", image_type);
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
-			return err;
+			return -EINVAL;
 		}
+
 		break;
 	}
 
 	case DOWNLOAD_CLIENT_EVT_DONE:
 		/* Write with 0 length to flush the write operation to flash. */
-		err = flash_img_buffered_write(&flash_img,
-				(u8_t *)event->fragment.buf,
-				0, true);
-		if (err != 0) {
-			LOG_ERR("flash_img_buffered_write error %d", err);
-			callback(FOTA_DOWNLOAD_EVT_ERROR);
-			return err;
-		}
+		if (image_type == MCUBOOT)  {
+			err = flash_img_buffered_write(&flash_img,
+						(u8_t *)event->fragment.buf,
+						0, true);
+			if (err != 0) {
+				LOG_ERR("flash_img_buffered_write error %d",
+					err);
+				callback(FOTA_DOWNLOAD_EVT_ERROR);
+				return err;
+			}
 
-		err = boot_request_upgrade(BOOT_UPGRADE_TEST);
-		if (err != 0) {
-			LOG_ERR("boot_request_upgrade error %d", err);
-			callback(FOTA_DOWNLOAD_EVT_ERROR);
-			return err;
+			err = boot_request_upgrade(BOOT_UPGRADE_TEST);
+			if (err != 0) {
+				LOG_ERR("boot_request_upgrade error %d", err);
+				callback(FOTA_DOWNLOAD_EVT_ERROR);
+				return err;
+			}
+		}
+#ifdef CONFIG_FOTA_DOWNLOAD_MODEM_FW_UPDATE_SUPPORT
+		else if (image_type == MODEM_FW) {
+			modem_fw_update();
```

Comment:
It should at least be logged in case we failed to schedule an update.

Reply:
Inapplicable won't fix.